server/handlers/pattern_collab_handler.go

```diff
@@ -0,0 +1,255 @@
+				// The hub closed the channel.
+				_ = c.conn.WriteMessage(websocket.CloseMessage, []byte{})
+				return
+			}
+
+			err := c.conn.SetWriteDeadline(time.Now().Add(writeWait))
+			if err != nil {
+				return
+			}
+
+			w, err := c.conn.NextWriter(websocket.TextMessage)
+			if err != nil {
+				return
+			}
+			_, err = w.Write(message)
+			if err != nil {
+				return
+			}
+
+			// Add queued chat messages to the current websocket message.
+			n := len(c.send)
+			for i := 0; i < n; i++ {
+				_, err = w.Write(newline)
+				if err != nil {
+					return
+				}
+
+				_, err = w.Write(<-c.send)
+				if err != nil {
+					return
+				}
+			}
+
+			if err := w.Close(); err != nil {
+				return
+			}
+		case <-ticker.C:
+			err := c.conn.SetWriteDeadline(time.Now().Add(writeWait))
+			if err != nil {
+				return
+			}
+			if err := c.conn.WriteMessage(websocket.PingMessage, nil); err != nil {
+				return
+			}
+		}
+	}
+}
+
+func (h *Hub) unsubscribeClientFromPattern(patternID string, client *Client) {
+	if h.patterns[patternID] != nil {
+		delete(h.patterns[patternID], client)
+		if len(h.patterns[patternID]) == 0 {
+			delete(h.patterns, patternID)
+		}
+	}
+}
+
+func (h *Hub) subscribeClientToPattern(patternID string, client *Client) {
+	if h.patterns[patternID] == nil {
+		h.patterns[patternID] = make(map[*Client]bool)
+	}
+	h.patterns[patternID][client] = true
+}
+
+func (h *Hub) closeChannel(ch chan []byte) {
+	select {
+	case <-ch:
+		return
+	default:
+	}
+	close(ch)
+}
+
+func (h *Hub) run() {
+	for {
+		select {
+		case client := <-h.register:
+			h.subscribeClientToPattern(client.patternID, client)
+		case client := <-h.unregister:
+			h.unsubscribeClientFromPattern(client.patternID, client)
+			client.connected = false
+			h.closeChannel(client.send)
+		case patternUpdateRequest := <-h.updateSubscribedPattern:
+			h.unsubscribeClientFromPattern(patternUpdateRequest.oldPatternID, patternUpdateRequest.client)
+			h.subscribeClientToPattern(patternUpdateRequest.newPatternID, patternUpdateRequest.client)
+		case message := <-h.broadcast:
+			if h.patterns[message.Topic] != nil {
+				msg, err := json.Marshal(message)
+				if err == nil {
+					for client := range h.patterns[message.Topic] {
+						if client.connected {
+							client.send <- msg
+						} else {
+							delete(h.patterns[message.Topic], client)
+						}
+					}
+				}
+			}
+		}
+	}
+}
+
+func newHub() *Hub {
+	return &Hub{
+		patterns:                make(map[string]map[*Client]bool),
+		broadcast:               make(chan *SocketMessage),
+		register:                make(chan *Client),
+		unregister:              make(chan *Client),
+		updateSubscribedPattern: make(chan *PatternUpdateRequest),
+	}
+}
+
+var mainHub *Hub = nil
+
+func serveWs(hub *Hub, w http.ResponseWriter, r *http.Request) {
+	conn, err := upgrader.Upgrade(w, r, nil)
+	if err != nil {
+		return
+	}
+
+	client := &Client{hub: hub, conn: conn, patternID: "", send: make(chan []byte), connected: true}
```

Comment:
why is patternid empty here?

Reply:
It is later sent by client when the type of message is "subscribe".